### PR TITLE
Add App Insights to base workspace

### PR DESCRIPTION
**Pull request by stuartleeks**
# PR for issue #1057

## What is being addressed

Add App Insights and Log Analytics to base workspace

## How is this addressed

- Add private DNS zones for Azure Monitor to base workspace
- Add App Insights and Log Analytics to base workspace

(case change in app_insights.json due to https://github.com/hashicorp/terraform-provider-azurerm/issues/17703)

### CODE REVIEW

File: templates/workspaces/base/terraform/azure-monitor/app_insights.json [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      },
+      "dependsOn": [
+        "[resourceId('microsoft.insights/components', parameters('app_insights_name'))]"
+      ]
+    }
+  ],
+  "outputs": {
+    "instrumentationKey": {
+      "value": "[reference(resourceId('microsoft.insights/components', parameters('app_insights_name')), '2020-02-02').InstrumentationKey]",
+      "type": "String"
```
Comment (tamirkamara) [inline templates/workspaces/base/terraform/azure-monitor/app_insights.json:60]: same here

File: templates/workspaces/base/terraform/azure-monitor/locals.tf [hunk 2]
```diff
@@ -0,0 +1,8 @@
+locals {
+  short_workspace_id = substr(var.tre_resource_id, -4, -1)
+  app_insights_name  = "appi-${var.tre_id}-ws-${local.short_workspace_id}"
+  tre_core_tags = {
+    tre_id              = var.tre_id
+    tre_core_service_id = var.tre_id
+  }
```
Comment (tamirkamara) [inline templates/workspaces/base/terraform/azure-monitor/locals.tf:7]: This doesn't look right. Should be a local from the root module named tre_workspace_tags with different set of values.
There're other references for this in other files.

File: templates/workspaces/base/terraform/azure-monitor/azure-monitor.tf [hunk 3]
```diff
@@ -0,0 +1,106 @@
+resource "azurerm_log_analytics_workspace" "core" {
```
Comment (tamirkamara) [inline templates/workspaces/base/terraform/azure-monitor/azure-monitor.tf:1]:
```suggestion
resource "azurerm_log_analytics_workspace" "workspace" {
```

Plus, creating a new log analytics workspace requires also setting up its customer provisioned storage account, doesn't it?
Something like this:
```terraform
resource "azurerm_log_analytics_linked_storage_account" "workspace_storage_ingestion" {
  data_source_type      = "Ingestion"
  resource_group_name   = var.resource_group_name
  workspace_resource_id = azurerm_log_analytics_workspace.core.id
  storage_account_ids   = [azurerm_storage_account.logs.id]
}

resource "azurerm_log_analytics_linked_storage_account" "workspace_storage_customlogs" {
  data_source_type      = "customlogs"
  resource_group_name   = var.resource_group_name
  workspace_resource_id = azurerm_log_analytics_workspace.core.id
  storage_account_ids   = [azurerm_storage_account.logs.id]
}
```

Comment (stuartleeks) [reply templates/workspaces/base/terraform/azure-monitor/azure-monitor.tf]: Added - thanks

**Comment by stuartleeks** [comment]
Current status:

PR build is seeing this error:

```2022-06-30 09:20:38 [    INFO] True, failed, cf1a78dd-083f-465d-be2a-be7b48ac37fb: Error context message = Unable to find image '***.azurecr.io/tre-workspace-base@sha256:8efb21cd80a81988362af07ade768b91448ab7a2c28e6053c098245c1e99313e' locally ╷│Error: Provider produced inconsistent result after apply││When applying changes to │module.azure_monitor.azurerm_monitor_private_link_scope.ampls, provider │"provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an │unexpected new value: Root resource was present, but now absent. ││This is a bug in the provider, which should be reported in the provider's │own issue tracker. ╵ Error: error running command /cnab/app/terraform terraform apply -auto-approve -input=false -var client_id=*** -var address_space=10.1.1.0/24 -var app_role_id_workspace_owner=3641b8b9-5212-4772-b298-3a5f999dc1fd -var app_role_id_workspace_researcher=cb876117-c0eb-429a-9ec2-0ff1940349d7 -var app_service_plan_sku=P1v3 -var auth_client_id=******* -var auth_client_secret=******* -var auth_tenant_id=******* -var aad_redirect_uris_b64=W10= -var client_secret=*** -var enable_local_debugging=false -var location=*** -var register_aad_application=false -var scope_id=api://*** -var shared_storage_quota=50 -var sp_id=0616d41c-d245-40df-aa54-5f223e983edf -var tre_id=*** -var tre_resource_id=cf1a78dd-083f-465d-be2a-be7b48ac37fb -var workspace_owner_object_id=ade81c51-edee-4f17-ac8d-1becb3dd7352: exit status 1 Error: mixin execution failed: exit status 1 Error: 2 errors occurred: 	* container exit code: 1, message: <nil>. fetching outputs failed: error copying outputs from container: Error: No such container:path: 5ea111ad12d0432c2f3c5ec7e3d9037393a419084de37003d5359086c0d3b173:/cnab/app/outputs 	* required output app_role_id_workspace_owner is missing and has no default   az login --identity -u 59fe7111-9873-4ac6-a428-9fa2149e6667 && az acr login --name *** && porter install "cf1a78dd-083f-465d-be2a-be7b48ac37fb"  --reference ***.azurecr.io/tre-workspace-base:v0.3.12  --param address_space="10.1.1.0/24" --param app_role_id_workspace_owner="3641b8b9-5212-4772-b298-3a5f999dc1fd" --param app_role_id_workspace_researcher="cb876117-c0eb-429a-9ec2-0ff1940349d7" --param arm_use_msi="true" --param azure_location="***" --param client_id="***" --param client_secret="***" --param id="cf1a78dd-083f-465d-be2a-be7b48ac37fb" --param register_aad_application="False" --param scope_id="api://***" --param sp_id="0616d41c-d245-40df-aa54-5f223e983edf" --param tfstate_container_name="tfstate" --param tfstate_resource_group_name="rg-***-mgmt" --param tfstate_storage_account_name="***mgmt" --param tre_id="***" --param workspace_owner_object_id="ade81c51-edee-4f17-ac8d-1becb3dd7352" --allow-docker-host-access --force --cred ./vmss_porter/arm_auth_local_debugging.json --cred ./vmss_porter/aad_auth.json && porter show cf1a78dd-083f-465d-be2a-be7b48ac37fb (resource.py:74)
2022-06-30 09:20:38 [   ERROR] Failed to deploy status message: cf1a78dd-083f-465d-be2a-be7b48ac37fb: Error context message = Unable to find image '***.azurecr.io/tre-workspace-base@sha256:8efb21cd80a81988362af07ade768b91448ab7a2c28e6053c098245c1e99313e' locally ╷│Error: Provider produced inconsistent result after apply││When applying changes to │module.azure_monitor.azurerm_monitor_private_link_scope.ampls, provider │"provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an │unexpected new value: Root resource was present, but now absent. ││This is a bug in the provider, which should be reported in the provider's │own issue tracker. ╵ Error: error running command /cnab/app/terraform terraform apply -auto-approve -input=false -var client_id=*** -var address_space=10.1.1.0/24 -var app_role_id_workspace_owner=3641b8b9-5212-4772-b298-3a5f999dc1fd -var app_role_id_workspace_researcher=cb876117-c0eb-429a-9ec2-0ff1940349d7 -var app_service_plan_sku=P1v3 -var auth_client_id=******* -var auth_client_secret=******* -var auth_tenant_id=******* -var aad_redirect_uris_b64=W10= -var client_secret=*** -var enable_local_debugging=false -var location=*** -var register_aad_application=false -var scope_id=api://*** -var shared_storage_quota=50 -var sp_id=0616d41c-d245-40df-aa54-5f223e983edf -var tre_id=*** -var tre_resource_id=cf1a78dd-083f-465d-be2a-be7b48ac37fb -var workspace_owner_object_id=ade81c51-edee-4f17-ac8d-1becb3dd7352: exit status 1 Error: mixin execution failed: exit status 1 Error: 2 errors occurred: 	* container exit code: 1, message: <nil>. fetching outputs failed: error copying outputs from container: Error: No such container:path: 5ea111ad12d0432c2f3c5ec7e3d9037393a419084de37003d5359086c0d3b173:/cnab/app/outputs 	* required output app_role_id_workspace_owner is missing and has no default   az login --identity -u 59fe7111-9873-4ac6-a428-9fa2149e6667 && az acr login --name *** && porter install "cf1a78dd-083f-465d-be2a-be7b48ac37fb"  --reference ***.azurecr.io/tre-workspace-base:v0.3.12  --param address_space="10.1.1.0/24" --param app_role_id_workspace_owner="3641b8b9-5212-4772-b298-3a5f999dc1fd" --param app_role_id_workspace_researcher="cb876117-c0eb-429a-9ec2-0ff1940349d7" --param arm_use_msi="true" --param azure_location="***" --param client_id="***" --param client_secret="***" --param id="cf1a78dd-083f-465d-be2a-be7b48ac37fb" --param register_aad_application="False" --param scope_id="api://***" --param sp_id="0616d41c-d245-40df-aa54-5f223e983edf" --param tfstate_container_name="tfstate" --param tfstate_resource_group_name="rg-***-mgmt" --param tfstate_storage_account_name="***mgmt" --param tre_id="***" --param workspace_owner_object_id="ade81c51-edee-4f17-ac8d-1becb3dd7352" --allow-docker-host-access --force --cred ./vmss_porter/arm_auth_local_debugging.json --cred ./vmss_porter/aad_auth.json && porter show cf1a78dd-083f-465d-be2a-be7b48ac37fb (resource.py:78)
```


I'm attempting to run the e2e extended tests in my own environment with the aim of adding extra debugging on the terraform to  narrow down the error. Hopefully this will either allow me to see where the error is and potentially update the Terraform config or gather more info for raising a bug on the azurerm provider

**Comment by stuartleeks** [comment]
/test-extended

(see whether the previous error is repeated:

**Comment by stuartleeks** [comment]
/test-extended

Retry after `A hash conflict was encountered for the role Assignment ID. Please use a new Guid.`

**Comment by stuartleeks** [comment]
/test-extended

Retry with increased timeout on test_create_guacamole_service_into_base_workspace to check if it passes

**Comment by stuartleeks** [comment]
/test-extended

Re-running due to eventgrid issue in last run

**Comment by stuartleeks** [comment]
/test-extended

re-running as previous run failed with `could not connect to registry.terraform.io`

**Comment by stuartleeks** [comment]
/test-extended

(previous run had ACR connectivity issue)

**Review by tamirkamara — APPROVED** [review]

**Comment by stuartleeks** [comment]
/test-force-approve

two previous successful runs

**Pull request merged**